presto-main/src/main/java/com/facebook/presto/execution/executor/MultilevelSplitQueue.java

```diff
@@ -0,0 +1,322 @@
+        checkState(result != null, "pollSplit cannot return null");
+
+        return result;
+    }
+
+    /**
+     * Normalizes the scheduled time counters for empty levels.
+     *
+     * If only one level has waiting splits, we use it's current scheduled time value to compute the
+     * expected time for other levels.
+     *
+     * If multiple levels have waiting splits, the expected scheduled time is computed based on the
+     * scheduled time of the level that deviates most from the expected value
+     *
+     * @return expected scheduled time for level 0
+     */
+    @GuardedBy("lock")
+    private long updateLevelTimes()
+    {
+        long level0ExpectedTime = levelScheduledTime[0];
+        boolean updated;
+        do {
+            double currentMultiplier = levelPriorityMultiplier;
+            updated = false;
+            for (int level = 0; level < LEVEL_THRESHOLD_SECONDS.length; level++) {
+                currentMultiplier /= levelPriorityMultiplier;
+                long levelExpectedTime = (long) (level0ExpectedTime * currentMultiplier);
+
+                if (levelWaitingSplits.get(level).isEmpty()) {
+                    levelScheduledTime[level] = levelExpectedTime;
+                    continue;
+                }
+
+                if (levelScheduledTime[level] > levelExpectedTime) {
+                    level0ExpectedTime = levelScheduledTime[level] * (long) Math.pow(levelPriorityMultiplier, level);
+                    updated = true;
+                    break;
+                }
+            }
+        } while (updated && level0ExpectedTime != 0);
+
+        return level0ExpectedTime;
+    }
+
+    @GuardedBy("lock")
+    private PrioritizedSplitRunner pollFirstSplit()
+    {
+        for (PriorityQueue<PrioritizedSplitRunner> level : levelWaitingSplits) {
+            PrioritizedSplitRunner split = level.poll();
+            if (split != null) {
+                return split;
+            }
+        }
+
+        return null;
+    }
+
+    public Priority updatePriority(Priority oldPriority, long quantaNanos, long scheduledNanos)
+    {
+        int oldLevel = oldPriority.getLevel();
+        int newLevel = computeLevel(scheduledNanos);
+
+        // prevents a misbehaving task from unfairly penalizing all other tasks in its level
+        long levelContribution = Math.min(quantaNanos, LEVEL_CONTRIBUTION_CAP);
+
+        if (oldLevel == newLevel) {
+            addLevelTime(oldLevel, levelContribution);
+            return new Priority(oldLevel, oldPriority.getLevelPriority() + quantaNanos);
+        }
+
+        long remainingLevelContribution = levelContribution;
+        long remainingTaskTime = quantaNanos;
+
+        // account time to intermediate levels properly
+        for (int currentLevel = oldLevel; currentLevel < newLevel; currentLevel++) {
+            // a level cannot accrue more time than the threshold gap between itself and the previous level
```

Comment:
I think it would be more clear to explain as something like
```
A task normally, accrues small runs in a level and then moves to the next, but if the task had a long run, accrue the max to each level as if it had run in that level to level time limit.